fxa/oauth.py

```diff
@@ -239,12 +254,18 @@ def verify_token(self, token, scope=None):
             resp = None
 
         if resp is None:
-            url = '/verify'
-            body = {
-                'token': token
-            }
-            resp = self.apiclient.post(url, body)
+            ## TODO we want to fetch https://oauth.accounts.firefox.com/.well-known/openid-configuration and then get the jwks_uri key to get the /v1/jwks url, but we'll just hardcosde it like this for now until we are closer to production. We need to do this because oauth.accounts.firefox.com does not yet respond to .well-known, but accounts.firefox.com does; for consistency, oauth.accounts.firefox.com should respond to .well-known.
 
+            keys = self.apiclient.get('/jwks').get('keys', [])
+            resp = None
+            for key in keys:
+                try:
+                    resp = self._decode_key(key, token)
+                    break
+                except jwt.exceptions.DecodeError:
+                    continue
```

Comment:
I think we'll benefit from a little more nuance around types of failure here. The tokenserver has [some special handling](https://github.com/mozilla-services/tokenserver/blob/f19ac0e8402b8203c1db51e9321ea50cb361f634/tokenserver/views.py#L220) for subclasses of `fxa.errors.Error` and in particular for the [`TrustError` subclass](https://github.com/mozilla-services/tokenserver/blob/f19ac0e8402b8203c1db51e9321ea50cb361f634/tokenserver/views.py#L226). To play nice with tokenserver's expectations, in my head this could look roughly like:

```
try:
    for key in keys:
        try:
            resp = self._decode_key(key, token)
            break
        except jwt.exceptions.InvalidSignatureError:
            # It's only worth trying other keys in the event of `InvalidSignature`; if it was invalid for
            # other reasons (e.g. it's expired) then using a different key won't help.
            # OTOH you might prefer to let the server do a bit of extra work just to keep the control-flow simpler.
            continue
except (jwt.exceptions.DecodeError, jwt.exceptions.InvalidKeyError):
    # It wasn't a JWT at all, or it was signed using a key type we don't support.
    # Fall back to asking the FxA server to verify.
    pass
except jwt.exceptions.Error as e:
    # Any other JWT-related failure (e.g. expired token) can immediately surface as a trust error.
    raise fxa.errors.TrustError({"error": str(e)})
```

It would also be great to have tests for a couple of those different failure modes.

Reply:
👍